### PR TITLE
testscript: merge coverage from all test binary executions

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
Please do not review yet. I'm still iterating on this.

**Comment by mvdan** [comment]
Alright, this is ready for a first round of review. Note that this change lacks tests and docs, because I want to know if the overall design is sane before I polish it.

The main use case is testing tools which are executed indirectly, for example via 'go build -toolexec=mytool'. It is possible to set up
"mytool" in $PATH via the test binary thanks to `TESTSCRIPT_COMMAND`, but then those indirectly executed sub-processes do not count towards the final coverage profile.

It's fine if you think the above is a hack, or that testscript isn't designed for this. However, it nearly works with minimal effort, except for coverage profiles.

The basic idea with this feature is that, if the test package's `TestMain` exports `TESTSCRIPT_COVER_EXTRA`, any coverage profiles found there after all scripts have run are merged into the final profile. A working example is probably the easiest way to show how this works: https://github.com/mvdan/garble/commit/9306f01f3debe51b5c47f1ee7fa1dba3c029afbf

Another way to design this API would be to pass some sort of parameter to `testscript.RunMain`. That seemed hard to do in a backwards-compatible way, though.

I did not try to design this API on a per-script basis since it makes little sense; `go test -coverprofile` works on all tests being run. Custom code in `TestMain` is perhaps a bit low-level, but reusing the test binary with `testscript` for direct calls to `$PATH` via `-toolexec` is already a pretty exotic use case that requires custom code. So I think it's fine for the API to be a bit rough, as long as it works :)

I also exposed a way for `TestMain` to tell if coverage profiles are being collected, to avoid that extra work if they are not.

**Comment by mvdan** [comment]
We discussed this offline, and came up with a new design that I'll implement over the next week or so.

File: testscript/testscript.go
Comment (rogpeppe) [inline testscript/testscript.go:321]: ... or just put it in the `Vars` initializer above (it doesn't matter if it's empty).

File: gotooltest/testdata/cover.txt
Comment (rogpeppe) [inline gotooltest/testdata/cover.txt:4]:
```suggestion
# Use the checked out module, based on where the test binary ran.
```

File: gotooltest/script_test.go
Comment (rogpeppe) [inline gotooltest/script_test.go:20]: Useful comment, thanks.

File: cmd/testscript/testdata/nogo.txt
Comment (rogpeppe) [inline cmd/testscript/testdata/nogo.txt:4]: How about a comment here?

```
# We can't just set PATH to empty
# because we need the part of it that
# contains the executable names, so use
# a special builtin instead.
```
?

**Review by rogpeppe — APPROVED** [review]
This looks great - it's actually simpler than I expected and should "just work" for people.
LGTM modulo a few comments and suggestions.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:62]: I didn't do that because I hoped to just reuse WorkdirRoot. But as far as I know that gets created later. If you confirm it's not possible to reuse that temp dir here, then yeah, I'll set up both under the same extra temp dir.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:104]: Sure. I don't think it matte

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:131]: Why do that, when we have TempFile? :) It is basically random numbers, but it also guards against (unlikely) collisions.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:149]: :shrug: The person who contributed this code for my project did it this way for Windows. I presume because executable bits aren't a thing there.

I don't have a problem trying harder with OpenFile, but bear in mind that I don't think it would make a difference and I don't have a way to test it either. All I can say is the current code has worked so far :)

File: testscript/testscript_test.go
Comment (mvdan) [reply testscript/testscript_test.go:63]: some tests want to `exec echo` the real /bin/echo, so there was a conflict with this `echo [file] [args...]` command.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:62]: Done the above.

File: testscript/testscript_test.go
Comment (mvdan) [reply testscript/testscript_test.go:63]: The commit message will mention this.

**Comment by mvdan** [comment]
Thanks for your review! I've applied nearly all your suggestions, minus two - I've left those threads unresolved.

I've also added a long commit message that summarizes the whole refactor.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go]: Funnily enough, I had to end up doing this, because otherwise I found tests out in the wild that would error due to empty profiles to be merged. That entire problem goes away if we don't create this file ourselves, and leave it to -test.coverprofile. The new comment explains it.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go]: I *also* had to end up doing this, because -toolexec goes past symlinks apparently. I also added a comment about this.

File: testscript/exe.go
Comment (rogpeppe) [reply testscript/exe.go]: for the record, you might not be using `go test` directly (you might've used `go test -c`)

File: testscript/exe.go
Comment (rogpeppe) [reply testscript/exe.go]: ah, cool!

File: testscript/exe.go
Comment (rogpeppe) [inline testscript/exe.go:152]: 32 bits isn't _that_ huge - might as well use 6 or 8 and make it as secure as a normal hash collision.

File: testscript/exe.go
Comment (rogpeppe) [inline testscript/exe.go:170]: tbh if you need O_TRUNC here, something is going wrong...

**Review by rogpeppe — APPROVED** [review]
LGTM thanks! (with a couple of trivial final suggestions)

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:170]: That was just a copy-paste from os.Create :) I don't think it matters either way, so I'll remove it.

File: testscript/exe.go
Comment (mvdan) [reply testscript/exe.go:152]: I actually started at 8, but went down a bit as I thought it was unnecessarily long :) I'll go with 6.

**Pull request merged**
